mllib/src/main/scala/org/apache/spark/mllib/clustering/AffinityPropagation.scala

```diff
@@ -0,0 +1,347 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.clustering
+
+import scala.collection.mutable.Map
+import scala.collection.mutable.Set
+
+import org.apache.spark.{Logging, SparkException}
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.graphx._
+import org.apache.spark.graphx.impl.GraphImpl
+import org.apache.spark.rdd.RDD
+
+/**
+ * :: Experimental ::
+ *
+ * Model produced by [[AffinityPropagation]].
+ *
+ * @param clusters the vertexIDs of each cluster.
+ * @param exemplars the vertexIDs of all exemplars.
+ */
+@Experimental
+class AffinityPropagationModel(
+    val clusters: Seq[Set[Long]],
+    val exemplars: Seq[Long]) extends Serializable {
+
+  /**
+   * Set the number of clusters
+   */
+  def getK(): Int = clusters.size
+
+  /**
+   * Find the cluster the given vertex belongs
+   */
+  def findCluster(vertexID: Long): Set[Long] = {
+    clusters.filter(_.contains(vertexID))(0)
+  } 
+ 
+  /**
+   * Find the cluster id the given vertex belongs
+   */
+  def findClusterID(vertexID: Long): Option[Int] = {
+    var i = 0
+    clusters.foreach(cluster => {
+      if (cluster.contains(vertexID)) {
+        return Some(i)
+      }
+      i += i
+    })
+    None 
+  } 
+}
+
+/**
+ * :: Experimental ::
+ *
+ * AffinityPropagation (AP), a graph clustering algorithm based on the concept of "message passing"
+ * between data points. Unlike clustering algorithms such as k-means or k-medoids, AP does not
+ * require the number of clusters to be determined or estimated before running it. AP is developed
+ * by [[http://www.psi.toronto.edu/affinitypropagation/FreyDueckScience07.pdf Frey and Dueck]].
+ *
+ * @param maxIterations Maximum number of iterations of the AP algorithm.
+ *
+ * @see [[http://en.wikipedia.org/wiki/Affinity_propagation (Wikipedia)]]
+ */
+@Experimental
+class AffinityPropagation private[clustering] (
+    private var maxIterations: Int,
+    private var lambda: Double,
+    private var normalization: Boolean) extends Serializable {
+
+  import org.apache.spark.mllib.clustering.AffinityPropagation._
+
+  /** Constructs a AP instance with default parameters: {maxIterations: 100, lambda: 0.5,
```

Comment:
Put backticks around `0.5`. Otherwise, ScalaDoc will only show "... lambda: 0." as the preview.